storage-node/src/services/sync/acceptPendingObjects.ts

```diff
@@ -54,61 +59,87 @@ export class AcceptPendingObjectsService {
+  private async processPendingObjects(pendingIds: string[]): Promise<PendingObjectDetails> {
     const pendingDataObjects = await this.qnApi.getDataObjectsByIds(pendingIds)
 
-    const deletedObjects = _.differenceWith(pendingIds, pendingDataObjects, (id, dataObject) => dataObject.id === id)
+    // objects not found in the query node
+    const maybeDeletedObjects = _.differenceWith(
+      pendingIds,
+      pendingDataObjects,
+      (id, dataObject) => dataObject.id === id
+    )
 
-    // Remove stale objects from the pending directory
-    if (deletedObjects.length) {
-      logger.warn(`Found data objects deleted from runtime in Pending directory: ${deletedObjects.length}`)
-      await Promise.all(deletedObjects.map((id) => fsPromises.unlink(path.join(this.pendingDataObjectsDir, id))))
+    if (maybeDeletedObjects.length) {
+      // unlink the pending object once we determined it is gone!
+      logger.debug(`Found ${maybeDeletedObjects.length} stale pending objects in pending folder`)
+      // Things to consider.
+      // These may be very recently uploaded files, note the uploads api uses chain rpc to
+      // validate upload of object that was created to allow uploads for objects that were created
+      // as soon as possible.
+      // const qnState = await this.qnApi.getQueryNodeState()
+      // get this.api chain head
+      // get timestamp in chain head block.. are we uptodate?
+      // test api.isSyncing?
+      // It would be easier to have query-node store object deleted events.
+      // if we chose a different file name convention for the pending objects,
+      // we could store the bag-id and object id in the filename to make it easy to lookup in runtime state.
     }
 
-    await Promise.all(
+    const objectsToAccept: PendingObjectDetails = []
+
+    await Promise.allSettled(
       pendingDataObjects.map(async (dataObject) => {
-        const storageBucket = dataObject.storageBag.storageBuckets.find(({ id }) => this.uploadBuckets.includes(id))
+        // verify the ipfshash of the object matches against runtime!
+        // just in-case wrong file was moved into pending folder manually
+        // ...
 
+        const storageBucket = dataObject.storageBag.storageBuckets.find(({ id }) => this.uploadBuckets.includes(id))
         if (storageBucket) {
           if (dataObject.isAccepted) {
             await this.movePendingDataObjectToUploadsDir(dataObject.id)
           } else {
-            this.add(dataObject.id, storageBucket.id, dataObject.storageBag.id)
+            objectsToAccept.push([dataObject.id, [storageBucket.id, dataObject.storageBag.id]])
           }
         } else {
-          logger.warn(
+          logger.debug(
             `Data object ${dataObject.id} in pending directory is no longer assigned to any of the upload buckets: ${this.uploadBuckets}.`
           )
+          await fsPromises.unlink(path.join(this.pendingDataObjectsDir, dataObject.id))
```

Comment:
I think we discussed that doing this is risky, and should probably by handled by pruning service (https://github.com/Joystream/joystream/pull/4971#discussion_r1409373556)

Reply:
Ah yes now I remember. That is a good idea actually.